File: src/domain/station/station.ts
import type { Branded } from "../units/branded";

/**
 * Un code station fait **dix** caractères (`"K447001001"`). Un code site en
 * fait huit (`"K4470010"`), et l'interroger renvoie chaque mesure **en double**
 * (C-05, vérifié le 2026-08-01). Le type interdit de confondre les deux.
 */
export type StationCode = Branded<string, "code_station">;

/**
 * Code de département INSEE, **en chaîne**. `"01"` traité comme un nombre
 * deviendrait `1` et ne correspondrait plus à rien côté API. La Corse
 * (`"2A"`, `"2B"`) le rendrait de toute façon impossible.
 */
export type DepartementCode = Branded<string, "code_departement">;

const STATION_CODE_LENGTH = 10;

/**
 * Forme **mesurée** sur les 4 150 stations en service le 2026-08-15, et non
 * supposée : 3 974 codes en `A999999999` (une lettre puis neuf chiffres) et
 * 176 en `9999999999` (dix chiffres) — ces derniers sont ceux des DOM.
 *
 * ⚠️ Restreindre à `^[A-Z]\d{9}$`, qui décrit pourtant la forme la plus
 * fréquente, rejetterait ces 176 stations. C'est exactement le genre de règle
 * qu'on n'écrit pas d'après un échantillon.
 */
const STATION_CODE_PATTERN = /^[A-Z0-9]{10}$/;
/** Deux chiffres, `2A`/`2B`, ou trois chiffres pour l'outre-mer. */
const DEPARTEMENT_PATTERN = /^(\d{2}|2[AB]|\d{3})$/;

export function stationCode(value: string): StationCode {
  if (value.length !== STATION_CODE_LENGTH) {
    throw new RangeError(
      `Un code station fait dix caractères, reçu ${value.length} : « ${value} ». ` +
        `Un code site (huit caractères) produirait des mesures en double (C-05).`,
    );
  }
  if (!STATION_CODE_PATTERN.test(value)) {
    throw new RangeError(
      `Forme de code station invalide : « ${value} ». Attendu dix caractères ` +
        `en majuscules et chiffres. Une valeur de la bonne longueur mais de la ` +
        `mauvaise forme partirait vers Hub'Eau et reviendrait vide, ce que ` +
        `l'écran afficherait en « pas de donnée » (BR-007).`,
    );
  }
  return value as StationCode;
}

export function departementCode(value: string): DepartementCode {
  const normalized = value.toUpperCase();
  if (!DEPARTEMENT_PATTERN.test(normalized)) {
    throw new RangeError(
      `Code de département attendu sur deux chiffres, « 2A »/« 2B », ou trois ` +
        `chiffres pour l'outre-mer. Reçu : « ${value} ».`,
    );
  }
  return normalized as DepartementCode;
}

/**
 * Station hydrométrique du référentiel Hub'Eau.
 *
 * ⚠️ Les coordonnées viennent de `latitude_station`/`longitude_station` dans le
 * référentiel, mais de `latitude`/`longitude` dans `observations_tr` — deux
 * noms pour la même chose, constaté le 2026-08-01. Le domaine n'en connaît
 * qu'un : c'est au mapper d'absorber l'écart.
 */
export interface Station {
  readonly code: StationCode;
  readonly libelle: string;
  readonly latitude: number;
  readonly longitude: number;
  readonly codeDepartement: DepartementCode;
  /** `null` quand le référentiel ne rattache la station à aucun cours d'eau. */
  readonly libelleCoursEau: string | null;
  readonly enService: boolean;
}

File: lib/domain/station/station.dart
// Le code station à dix caractères et le code département en chaîne (C-05).
// C-05 reproduit le 2026-09-13 : le code SITE K4470010 (8 car.) renvoie
// chaque mesure en double, dont une ligne à code_station null — count 430
// contre 216 pour le code station. Une valeur de bonne longueur mais de
// mauvaise forme partirait vers l'API et reviendrait vide, affichée en
// « pas de donnée » (BR-007) : un faux négatif silencieux. On refuse à la
// construction plutôt que de laisser passer.

/// Forme exacte des codes station en service, mesurée sur les 4 150
/// stations du référentiel le 2026-09-13 : 3 974 en `A999999999` (une
/// lettre puis neuf chiffres), 176 en `9999999999` (dix chiffres, DOM).
/// Restreindre à la seule forme majoritaire rejetterait ces 176 stations.
final RegExp _stationCodePattern = RegExp(r'^[A-Z0-9]{10}$');

/// Code station Hub'Eau, toujours à dix caractères. Une classe et non un
/// `extension type` : à la différence des unités de
/// `lib/domain/units/quantities.dart`, ce type **valide** — c'est le seul
/// verrou mécanique contre `C-05`. La casse n'est jamais normalisée : un
/// `k` minuscule est refusé, il ne correspond à aucune forme mesurée.
final class StationCode {
  /// Valide [raw] contre [_stationCodePattern]. Lève une [ArgumentError] si
  /// [raw] n'a pas exactement dix caractères, ou contient un caractère hors
  /// `A-Z0-9` (minuscule comprise).
  factory StationCode(String raw) {
    if (!_stationCodePattern.hasMatch(raw)) {
      throw ArgumentError.value(
        raw,
        'raw',
        'doit comporter dix caractères A-Z0-9 (code station, jamais le '
            'code site à huit caractères — C-05)',
      );
    }

    return StationCode._(raw);
  }

  const StationCode._(this.value);

  /// Le code, tel que validé — jamais transformé.
  final String value;

  @override
  bool operator ==(Object other) =>
      other is StationCode && other.value == value;

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => value;
}

/// Forme du code département : deux chiffres, `2A`/`2B` pour la Corse, ou
/// trois chiffres pour les DOM.
final RegExp _departementCodePattern = RegExp(r'^(\d{2}|2[AB]|\d{3})$');

/// Code département, en chaîne et non en entier : `"01"` interprété comme un
/// nombre deviendrait `1`, et la Corse (`2A`/`2B`) rendrait la conversion de
/// toute façon impossible.
final class DepartementCode {
  /// Normalise [raw] (`trim` puis majuscules), puis le valide contre
  /// [_departementCodePattern]. Lève une [ArgumentError] sinon.
  factory DepartementCode(String raw) {
    final String normalized = raw.trim().toUpperCase();
    if (!_departementCodePattern.hasMatch(normalized)) {
      throw ArgumentError.value(
        raw,
        'raw',
        'doit être un code département valide : deux chiffres, 2A/2B, ou '
            'trois chiffres pour les DOM',
      );
    }

    return DepartementCode._(normalized);
  }

  const DepartementCode._(this.value);

  /// Le code, normalisé en majuscules.
  final String value;

  @override
  bool operator ==(Object other) =>
      other is DepartementCode && other.value == value;

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => value;
}

/// Une station du référentiel Hub'Eau hydrométrie.
///
/// Les coordonnées s'appellent `latitude_station`/`longitude_station` au
/// référentiel et `latitude`/`longitude` en temps réel : le domaine n'en
/// connaît qu'un couple, c'est au mapper d'absorber cet écart de
/// nomenclature de l'API.
final class Station {
  const Station({
    required this.code,
    required this.label,
    required this.latitude,
    required this.longitude,
    required this.departement,
    required this.riverLabel,
    required this.inService,
  });

  /// Code de la station, à dix caractères.
  final StationCode code;

  /// Libellé de la station.
  final String label;

  final double latitude;

  final double longitude;

  /// Département où se situe la station.
  final DepartementCode departement;

  /// Cours d'eau de la station. `null` signifie une absence — jamais une
  /// chaîne vide (`BR-007`).
  final String? riverLabel;

  /// `true` si la station est en service.
  final bool inService;
}
